src/mappings/event/tokens.ts

```diff
@@ -0,0 +1,461 @@
+import { SubstrateBlock } from "@subsquid/substrate-processor";
+import { Ctx, EventItem, getCirculatingSupplyProcessRange } from "../../processor";
+import {
+    TokensReservedEvent,
+    TokensUnreservedEvent,
+    TokensLockedEvent,
+    TokensUnlockedEvent,
+    TokensDepositedEvent,
+    TokensWithdrawnEvent,
+    TokensTransferEvent
+} from "../../types/events";
+import { CurrencyId as CurrencyId_V1 } from "../../types/v1";
+import { CurrencyId as CurrencyId_V6 } from "../../types/v6";
+import { CurrencyId as CurrencyId_V10 } from "../../types/v10";
+import { CurrencyId as CurrencyId_V15 } from "../../types/v15";
+import { CurrencyId as CurrencyId_V17 } from "../../types/v17";
+import { CurrencyId as CurrencyId_V1020000 } from "../../types/v1020000";
+import { CurrencyId as CurrencyId_V1021000 } from "../../types/v1021000";
+import { CurrencyId as CurrencyId_V1024000 } from "../../types/v1024000";
+import EntityBuffer from "../utils/entityBuffer";
+import { blockToHeight } from "../utils/heights";
+import { getNativeCurrency } from "../utils/nativeCurrency";
+import { UpdateType, updateCumulativeCirculatingSupply } from "../utils/cumulativeCirculatingSupply";
+import { address, currencyId, isSystemAddress, legacyCurrencyId } from "../encoding";
+import { CumulativeCirculatingSupply, Currency, Token, Transfer } from "../../model";
+import { convertAmountToHuman } from "../_utils";
+
+function isSameCurrency(
+    expectedNativeCurrency: Token.INTR | Token.KINT,
+    currencyId: CurrencyId_V1 |
+        CurrencyId_V6 |
+        CurrencyId_V10 |
+        CurrencyId_V15 |
+        CurrencyId_V17 |
+        CurrencyId_V1020000 |
+        CurrencyId_V1021000
+) : boolean {
+    return currencyId.__kind === expectedNativeCurrency || 
+    (currencyId.__kind === "Token" && currencyId.value.__kind === expectedNativeCurrency);
+
+}
```

Comment:
I noticed that this is always called after the native currency is obtained, what do you think about getting the native currency inside this function:

```suggestion
function isSameCurrency(
    currencyId: CurrencyId_V1 |
        CurrencyId_V6 |
        CurrencyId_V10 |
        CurrencyId_V15 |
        CurrencyId_V17 |
        CurrencyId_V1020000 |
        CurrencyId_V1021000
) : boolean {
     const expectedNativeCurrency = getNativeCurrency();
    return currencyId.__kind === expectedNativeCurrency || 
    (currencyId.__kind === "Token" && currencyId.value.__kind === expectedNativeCurrency);

}
```

Reply:
> I noticed that this is always called after the native currency is obtained, what do you think about getting the native currency inside this function

That's true, but we also need to keep the native currency anyway because it is passed through as param for `updateCumulativeCirculatingSupply(...)`
So, rather than calling `getNativeCurrency()` twice once inside the currency check method and then once again to pass through to `updateCumulativeCirculatingSupply(...)`, I opted to just keep it from the first call.